File: leetcode/two-sum.cpp
/* -------------------------------------------------------------------------- */
/*                   https://leetcode.com/problems/two-sum/                   */
/* -------------------------------------------------------------------------- */

#include <bits/stdc++.h>

using namespace std;

class Solution {
   public:
    vector<int> FindTwoSumValues(vector<int> nums, int target) {
        std::sort(nums.begin(), nums.end());
        int i = 0;
        int j = nums.size() - 1;
        while (i < j) {
            if (nums[i] + nums[j] == target) {
                return {nums[i], nums[j]};
            }
            if (nums[i] + nums[j] < target) {
                i++;
            } else {
                j--;
            }
        }
        return {-1, -1};
    }

    vector<int> twoSum(vector<int>& nums, int target) {
        auto cands = FindTwoSumValues(nums, target);
        int i = -1, j = -1;
        for (int k = 0; k < nums.size(); k++) {
            if (i == -1 && nums[k] == cands[0]) {
                i = k;
                nums[k] = -1;
            }
            if (j == -1 && nums[k] == cands[1]) {
                j = k;
            }
        }
        return {i, j};
    }
};

int main() {
    vector<int> nums = {2, 7, 11, 15};
    int target = 9;

    auto ans = Solution().twoSum(nums, target);
    std::cout << ans[0] << " " << ans[1] << "\n";

    nums = {3, 2, 4};
    target = 6;
    ans = Solution().twoSum(nums, target);
    std::cout << ans[0] << " " << ans[1] << "\n";

    nums = {3, 3};
    target = 6;
    ans = Solution().twoSum(nums, target);
    std::cout << ans[0] << " " << ans[1] << "\n";
}

File: leetcode/two-sum.rs
/* -------------------------------------------------------------------------- */
/*                   https://leetcode.com/problems/two-sum/                   */
/* -------------------------------------------------------------------------- */

pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut i = 0;
    let mut j = nums.len() - 1;

    if nums.len() == 2 {
        return vec![0, 1];
    }

    let mut sorted_vec = Vec::with_capacity(nums.len());
    for (idx, val) in nums.iter().enumerate() {
        sorted_vec.push((val, idx as i32));
    }

    sorted_vec.sort_by_key(|k| k.0);
    while i < j {
        let sum = sorted_vec[i].0 + sorted_vec[j].0;
        if sum == target {
            return vec![sorted_vec[i].1, sorted_vec[j].1];
        } else if sum < target {
            i += 1;
        } else {
            j -= 1;
        }
    }

    return vec![];
}

fn main() {
    println!("{:?}", two_sum(vec![3, 2, 4], 6));
}
